Issue 2244

trendels opened the pull request:
This PR contains a test and fix for #2244 

The "chicken and egg problem" with the (inotify-based) reloader depending on the wsgi app to be loaded first, and `load_wsgi` depending on the reloader to be present is solved by initializing the reloader first, but only starting it after the wsgi app was loaded.

I've added an integration test that runs `gunicorn --reload` in a tmpdir, replaces the WSGI app and checks for changes in the HTTP response via a unix domain socket. In order to not slow down the tests too much, I've made the default poll interval for the poll-based reloader (1 second) overridable via an (undocumented) [environment variable](https://github.com/benoitc/gunicorn/pull/2249/commits/f1f0d011aec16adcb7b28b7fb5b0f3bb391de2fe). I hope this is OK.

Initially, the test was very flaky and failed most of the time even with the fix (but only when running under `pytest`): gunicorn would pick up the changed file and restart the worker, but the new worker would still be running the old code! It took me a while to figure out that this must have something to do with the compiled bytecode (maybe a timestamp issue?). [Starting gunicorn with writing of bytecode disabled](https://github.com/benoitc/gunicorn/pull/2249/commits/6923709396afca7f8966bc19528094bd5af566d0#diff-3d4ef4fd9b978e57e60dfb39ca4c5baeR27-R31) fixed the issue for me. I was not able to reproduce this outside of pytest.

There was also an issue where the inotify-based reloader would throw an exception if a directory it tried to watch did not exist, which happens when using `--reload` and `--chdir` together (the poll-based reloader already ignores all missing/unreadable files).

trendels commented:
I see there are still some failures on Travis CI. I'm not sure what the issue is. Possibly the same one I was seeing earlier. The tests pass when I run them locally (except pypy which I don't have installed):

```
$ python3 -m venv tox-venv
$ tox-venv/bin/pip install --upgrade tox coverage
$ tox-venv/bin/tox --recreate
[...]
--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)

________________________________________________ summary ________________________________________________
  py35: commands succeeded
  py36: commands succeeded
  py37: commands succeeded
  py38: commands succeeded
ERROR:  pypy3: InterpreterNotFound: pypy3
  lint: commands succeeded
```

trendels commented:
I've ran the tests in CI some more, and they [remain flaky unfortunately](https://travis-ci.org/trendels/gunicorn/builds), even with slightly increased timeouts.

I don't think the failures are real errors, but that the detection of filesystem changes can be arbitrarily slow in the CI environment. Maybe the best way would be to disable this test in CI?

tilgovi commented on gunicorn/workers/base.py:
```diff
@@ -132,7 +130,9 @@ def changed(fname):
-            self.reloader.start()
+
+        self.load_wsgi()
+        self.reloader.start()
```
This needs to be wrapped in a conditional in case the reloader doesn't exist, right?

trendels replied:
You're right, I've fixed it.

tilgovi commented on gunicorn/reloader.py:
```diff
@@ -12,10 +12,11 @@
 import threading
 
 COMPILED_EXT_RE = re.compile(r'py[co]$')
+POLL_INTERVAL = float(os.getenv('GUNICORN_RELOADER_POLL_INTERVAL', '1'))
```
Should we expose this on the CLI somehow? Or is this intended just for the tests?

trendels replied:
This is just intended for the tests, so the test does not take multiple seconds each time.

tilgovi commented:
If you'd like to squash up the commits, feel free to do so!

berkerpeksag commented on gunicorn/workers/base.py:
```diff
@@ -132,6 +130,9 @@ def changed(fname):
                                          callback=changed)
+
+        self.load_wsgi()
+        if self.reloader:
```
Why do we need this branch? Can't we just simply revert https://github.com/benoitc/gunicorn/commit/3054b184dea1b07581293c203a1c546c7ea3ebca?

trendels replied:
That commit was meant as a fix for #2133 (inotify reloader does not pick up files), but also introduced the regression in  #2244 (syntax errors crash the process on reload). So a simple revert would fix the regression but also re-break the inotify reloader unfortunately.

Doing things in this particular order (load_wsgi, then start the reloader) fixes the original problem, but without the regression.

trendels commented:
> I wonder if it would be possible to add a test case for original problem that broke reloading of files with syntax error? See #2133 for the original report.

The cause for this was an uninitialized variable, so I guess I can add a unit test for that.

I suggest that I do that and remove the integration test instead. While I think it would be very useful to have such a test, I continue to see spurious failures when running it, and I don't want to introduce a flaky test to the test suite.

trendels commented:
I've replaced the flaky test with more targeted unit tests for the specific issues #2244 and #2133 that this PR addresses, and squashed everything into two commits (two because f39b065 fixes an unrelated issue that I noticed during testing).

tilgovi commented:
This is great, @trendels. Thanks for being responsive on the tests and feedback. Feel free to nudge me for followup if I'm being slow.
